File: Viewer/src/DTOs/StageViewDTO.js
import React from 'react';

export default class StageViewDTO{
    constructor(type, numCompetitorsIn, numCompetitorsOut, numCompetitorsPerMatch, numLegs, matches, numTeamsPerGroup, pointsForWin, pointsForTie, pointsForLoss) {
        this.type = type;
        this.numCompetitorsIn = numCompetitorsIn;
        this.numCompetitorsOut = numCompetitorsOut;
        this.numCompetitorsPerMatch = numCompetitorsPerMatch;
        this.numLegs = numLegs;
        this.matches = matches;
        this.numTeamsPerGroup = numTeamsPerGroup;
        this.pointsForWin = pointsForWin;
        this.pointsForTie = pointsForTie;
        this.pointsForLoss = pointsForLoss;
    }
    getState(){
        switch (this.type) {
            case "PLAY_OFF":
                return { type: this.type, numCompetitorsIn: this.numCompetitorsIn,
                    numCompetitorsOut: this.numCompetitorsOut, numLegs: this.numLegs}
            case "GROUP":
                return { type: this.type, numCompetitorsIn: this.numCompetitorsIn,
                    numCompetitorsOut: this.numCompetitorsOut, numLegs: this.numLegs,
                    numTeamsPerGroup: this.numTeamsPerGroup, pointsForWin: this.pointsForWin,
                    pointsForTie: this.pointsForTie, pointsForLoss: this.pointsForLoss}
            case "FREE_FOR_ALL":
                return { type: this.type, numCompetitorsIn: this.numCompetitorsIn,
                    numCompetitorsOut: this.numCompetitorsOut, numCompetitorsPerMatch: this.numCompetitorsPerMatch,
                    numLegs: this.numLegs}
            case "LEAGUE":
                return { type: this.type, numCompetitorsIn: this.numCompetitorsIn,
                    numCompetitorsOut: this.numCompetitorsOut, numLegs: this.numLegs,
                    pointsForWin: this.pointsForWin, pointsForTie: this.pointsForTie,
                    pointsForLoss: this.pointsForLoss}
            default:
                alert("problem, wrong stage type");
        }

    }
    // val matches: Set<MatchViewDto> = setOf()
}

File: backend/src/main/kotlin/com/esport/kompetitor/persistence/dto/competition/StageViewDto.kt
package com.esport.kompetitor.persistence.dto.competition

import com.esport.kompetitor.persistence.entity.*

data class StageViewDto(
    val id: Long? = null,
    val type: Type,
    val numCompetitorsIn: Int,
    val numCompetitorsOut: Int,
    val numCompetitorsPerMatch: Int,
    val numLegs: Int,
    val matches: Set<MatchViewDto> = setOf(),
    val numTeamsPerGroup: Int? = null,
    val pointsForWin: Int? = null,
    val pointsForTie: Int? = null,
    val pointsForLoss: Int? = null,
    val groups: List<GroupViewDto>? = null,
) {
    companion object {
        enum class Type {
            LEAGUE, GROUP, PLAY_OFF, FREE_FOR_ALL
        }

        private fun fromLeague(league: LeagueStage) = league.run { StageViewDto(
            id = id,
            type = Type.LEAGUE,
            numCompetitorsIn = numCompetitorsIn,
            numCompetitorsOut = numCompetitorsOut,
            numCompetitorsPerMatch = numCompetitorsPerMatch,
            numLegs = numLegs,
            matches = matches.map { MatchViewDto.fromMatch(it) }.toSet(),
            pointsForWin = pointsForWin,
            pointsForTie = pointsForTie,
            pointsForLoss = pointsForLoss,
        ) }

        private fun fromGroupStage(group: GroupStage) = group.run { StageViewDto(
            id = id,
            type = Type.GROUP,
            numCompetitorsIn = numCompetitorsIn,
            numCompetitorsOut = numCompetitorsOut,
            numCompetitorsPerMatch = numCompetitorsPerMatch,
            numLegs = numLegs,
            matches = matches.map { MatchViewDto.fromMatch(it) }.toSet(),
            numTeamsPerGroup = numTeamsPerGroup,
            pointsForWin = pointsForWin,
            pointsForTie = pointsForTie,
            pointsForLoss = pointsForLoss,
            groups = groups.map { GroupViewDto.fromGroup(it) },
        ) }

        private fun fromPlayOff(playOff: PlayOffStage) = playOff.run { StageViewDto(
            id = id,
            type = Type.PLAY_OFF,
            numCompetitorsIn = numCompetitorsIn,
            numCompetitorsOut = numCompetitorsOut,
            numCompetitorsPerMatch = numCompetitorsPerMatch,
            numLegs = numLegs,
            matches = matches.map { MatchViewDto.fromMatch(it) }.toSet(),
            groups = groups.map { GroupViewDto.fromGroup(it) },
        ) }

        private fun fromFreeForAll(freeForAll: FreeForAllStage) = freeForAll.run { StageViewDto(
            id = id,
            type = Type.FREE_FOR_ALL,
            numCompetitorsIn = numCompetitorsIn,
            numCompetitorsOut = numCompetitorsOut,
            numCompetitorsPerMatch = numCompetitorsPerMatch,
            numLegs = numLegs,
            matches = matches.map { MatchViewDto.fromMatch(it) }.toSet(),
        ) }

        fun fromStage(stage: Stage) = when(stage) {
            is LeagueStage -> fromLeague(stage)
            is GroupStage -> fromGroupStage(stage)
            is PlayOffStage -> fromPlayOff(stage)
            is FreeForAllStage -> fromFreeForAll(stage)
            else -> throw IllegalArgumentException("Unknown type: ${stage.javaClass}")
        }
    }

    fun toStage(competition: Competition): Stage {
        require(matches.isEmpty()) { "Only stages that have no matches yet should be constructed into objects." }
        return when(type) {
            Type.LEAGUE -> LeagueStage(
                competition = competition,
                competitorsIn = numCompetitorsIn,
                competitorsOut = numCompetitorsOut,
                numLegs = numLegs,
                pointsForWin = pointsForWin!!,
                pointsForTie = pointsForTie!!,
                pointsForLoss = pointsForLoss!!)
            Type.GROUP -> GroupStage(
                competition = competition,
                competitorsIn = numCompetitorsIn,
                competitorsOut = numCompetitorsOut,
                numLegs = numLegs,
                numTeamsPerGroup = numTeamsPerGroup!!,
                pointsForWin = pointsForWin!!,
                pointsForTie = pointsForTie!!,
                pointsForLoss = pointsForLoss!!)
            Type.PLAY_OFF -> PlayOffStage(
                competition = competition,
                competitorsIn = numCompetitorsIn,
                competitorsOut = numCompetitorsOut,
                numLegs = numLegs)
            Type.FREE_FOR_ALL -> FreeForAllStage(
                competition = competition,
                competitorsIn = numCompetitorsIn,
                competitorsOut = numCompetitorsOut,
                numCompetitorsPerMatch = numCompetitorsPerMatch,
                numLegs = numLegs)
        }
    }
}